File: LeetCode/Java/20-stack-有效的括号.java

/*
 * @lc app=leetcode.cn id=20 lang=java
 *
 * [20] 有效的括号
 */
import java.util.*;

// @lc code=start
class Solution {
    public boolean isValid(String s) {
        var stack = new Stack<Character>();
        for (char c : s.toCharArray()) {
            if (c == '(')
                stack.push(')');
            else if (c == '[')
                stack.push(']');
            else if (c == '{')
                stack.push('}');
            else if (stack.empty() || c != stack.pop())
                return false;
        }
        return stack.empty();
    }
}
// @lc code=end

File: LeetCode/Go/20-stack-有效的括号.go
/*
 * @lc app=leetcode.cn id=20 lang=golang
 *
 * [20] 有效的括号
 */
package leetcode

// @lc code=start
func isValid(s string) bool {
	var stack = []rune{}
	for _, c := range s {
		if c == ')' || c == ']' || c == '}' {
			if len(stack) == 0 || stack[len(stack)-1] != c {
				return false
			} else {
				stack = stack[:len(stack)-1]
			}
		} else if c == '(' {
			stack = append(stack, ')')
		} else if c == '[' {
			stack = append(stack, ']')
		} else if c == '{' {
			stack = append(stack, '}')
		}
	}
	return len(stack) == 0
}

// @lc code=end
